tests/base/test_array.py

```diff
@@ -0,0 +1,467 @@
+        lower_w = rng.normal(size=(4, 100))
+        upper_w = rng.normal(size=(4, 100))
+        result = array_stats.power_scale_sense(ary, lower_w, upper_w, -0.5, 0.5)
+        assert result.shape == ()
+
+
+class TestComputeRanks:
+    def test_compute_ranks_basic(self, array_stats):
+        x = np.array([3.0, 1.0, 2.0, 4.0])
+        result = array_stats.compute_ranks(x)
+        assert result.shape == (4,)
+
+    def test_compute_ranks_relative(self, array_stats):
+        x = np.array([3.0, 1.0, 2.0, 4.0])
+        result = array_stats.compute_ranks(x, relative=True)
+        assert np.all(result >= 0)
+        assert np.all(result <= 1)
+
+    @pytest.mark.parametrize("axis", [0, 1, -1])
+    def test_compute_ranks_axis(self, array_stats, axis):
+        rng = np.random.default_rng(42)
+        ary = rng.normal(size=(10, 20, 30))
+        result = array_stats.compute_ranks(ary, axis=axis)
+        assert result.ndim == ary.ndim
+
+
+class TestBinning:
+    def test_get_bins_basic(self, array_stats):
+        rng = np.random.default_rng(42)
+        ary = rng.normal(size=(100,))
+        bins = array_stats.get_bins(ary)
+        assert bins.ndim == 1
+        assert len(bins) > 0
+
+    @pytest.mark.parametrize("bins", [10, 20, 50])
+    def test_get_bins_integer(self, array_stats, bins):
+        rng = np.random.default_rng(42)
+        ary = rng.normal(size=(100,))
+        result = array_stats.get_bins(ary, bins=bins)
+        assert result.ndim == 1
+
+    def test_histogram_basic(self, array_stats):
+        rng = np.random.default_rng(42)
+        ary = rng.normal(size=(1000,))
+        counts, edges = array_stats.histogram(ary)
+        assert len(counts) == len(edges) - 1
+        assert counts.sum() == 1000
+
+    @pytest.mark.parametrize("bins", [10, 20, 50])
+    def test_histogram_bins(self, array_stats, bins):
+        rng = np.random.default_rng(42)
+        ary = rng.normal(size=(1000,))
+        counts, _ = array_stats.histogram(ary, bins=bins)
+        assert len(counts) == bins
+
+    def test_histogram_density(self, array_stats):
+        rng = np.random.default_rng(42)
+        ary = rng.normal(size=(1000,))
+        counts, edges = array_stats.histogram(ary, density=True)
+        widths = np.diff(edges)
+        integral = np.sum(counts * widths)
+        assert_allclose(integral, 1.0, rtol=0.01)
+
+    def test_histogram_range(self, array_stats):
+        rng = np.random.default_rng(42)
+        ary = rng.normal(size=(1000,))
+        _, edges = array_stats.histogram(ary, bins=20, range=(-2, 2))
+        assert_allclose(edges[0], -2)
+        assert_allclose(edges[-1], 2)
+
+
+class TestKDE:
+    def test_kde_basic(self, array_stats):
+        rng = np.random.default_rng(42)
+        ary = rng.normal(size=(100,))
+        grid, pdf, bw = array_stats.kde(ary)
+        assert len(grid) == 512
+        assert len(pdf) == 512
+        assert bw > 0
+
+    @pytest.mark.parametrize("grid_len", [128, 256, 512, 1024])
```

Comment:
I don't think this will actually take very long, but given the test nature, I think the first two are enough to check what we care about